### PR TITLE
2.0 API profiles routes

### PR DESCRIPTION
**Pull request by srinia6**
This PR has contributions from @srinia6 @geoff-reid @BillyAbildgaard @dalebremner and @brianparry .

Requires https://github.com/RackHD/on-http/pull/296 PR 
@RackHD/corecommitters @zyoung51 @jlongever @amymullins

### CODE REVIEW

File: spec/lib/api/2.0/profiles-spec.js [hunk 1]
```diff
@@ -0,0 +1,373 @@
...
+    describe('2.0 GET /profiles/metadata', function () {
+        it('should return a list of profiles', function () {
+            profiles.getAll.resolves(profile);
+            return helper.request().get('/api/2.0/profiles/metadata')
+                .expect('Content-Type', /^application\/json/)
+                .expect(200)
+                .then(function(res) {
+                    expect(res.body[0]).to.have.property('name', 'dummy_profile');
+                    expect(res.body[0]).to.have.property('scope', 'dummy');
+		            expect(res.body[0]).to.have.property('hash', '123');
```
Comment (yyscamper) [inline spec/lib/api/2.0/profiles-spec.js]: indent error

File: spec/lib/api/2.0/profiles-spec.js [hunk 2]
```diff
@@ -0,0 +1,373 @@
...
+		            expect(res.body[0]).to.not.have.property('blank');
```
Comment (yyscamper) [inline spec/lib/api/2.0/profiles-spec.js]: A little odd to me, why expect not have property `blank`? If you want to expect the response has no extra properties, you need to check all keys of res.body[0] to exactly equal to a list. such as:
```javascript
expect(_.keys(res.body[0])).to.deep.equal(['key1', 'key2', ...]);
```

File: spec/lib/api/2.0/profiles-spec.js [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+    });
+
+    describe('2.0 GET /profiles/metadata/:name', function () {
+        it('should return a single profile', function () {
+            profiles.getName.resolves(profile);
+            return helper.request().get('/api/2.0/profiles/metadata/dummy_profile')
+            .expect('Content-Type', /^application\/json/)
+            .expect(200)
+	        .then(function(res) {
+		        expect(res.body[0]).to.have.property('name', 'dummy_profile');
```
Comment (yyscamper) [inline spec/lib/api/2.0/profiles-spec.js]: indent error

File: spec/lib/api/2.0/profiles-spec.js [hunk 4]
```diff
@@ -0,0 +1,373 @@
...
+                    expect(profiles.get).to.have.been.calledWith('dummy_profile', 'dummy');
+            });
+        });
+
+        it('should return 404 on invalid profile name', function() {
+            profiles.get.rejects(new Errors.NotFoundError('invalid profile'));
+            return helper.request().get('/api/2.0/profiles/library/0000?scope=dummy')
+                .expect(404)
+                .then(function() {
+                expect(profiles.get).to.have.been.calledOnce;
```
Comment (yyscamper) [inline spec/lib/api/2.0/profiles-spec.js]: indent error

File: spec/lib/api/2.0/profiles-spec.js [hunk 5]
```diff
@@ -0,0 +1,373 @@
...
+                expect(profiles.get).to.have.been.calledWith('0000');
+                });
+        });
+    });
+
+    describe('2.0 PUT /profiles/library/:name', function () {
+        var temp;
+        beforeEach('should PUT new mockfile', function () {
```
Comment (yyscamper) [inline spec/lib/api/2.0/profiles-spec.js]: Seems this `beforeEach` should be `it`

File: spec/lib/api/2.0/profiles-spec.js [hunk 6]
```diff
@@ -0,0 +1,373 @@
...
+                .expect(function(){
+                    expect(profiles.put).to.have.been.calledOnce;
+                    expect(profiles.put).to.have.been.calledWith('test.ipxe');
+                });
+        });
+
+        it('should get new profile', function () {
+            return helper.request().get('/api/2.0/profiles/metadata/test.ipxe')
+            .expect(200)
+                .then(function (res) {
```
Comment (yyscamper) [inline spec/lib/api/2.0/profiles-spec.js]: indent error

**Comment by yyscamper** [comment]
@srinia6 : one more comment, please keep your unit-code in consistent style, I see mostly you write following code:
```javascript
return helper.request('xxx')
    .expect(xxx)
    .expect(yyy)
```
But sometime no indent for each `expect`.
```javascript
return helper.request('xxx')
.expect(xxx)
.expect(yyy)
```

Either style is OK, but need be in a consistent  style at least in the same file.

**Comment by yyscamper** [comment]
:+1: after fix all my comment and houndci warning.

File: spec/lib/api/2.0/profiles-spec.js (hunk 5 above)
Comment (srinia6) [inline spec/lib/api/2.0/profiles-spec.js]: @yyscamper Could be, i was trying to PUT new profile and do a GET.......

Comment (yyscamper) [inline spec/lib/api/2.0/profiles-spec.js]: @srinia6 : why not using stub?

**Comment by tannoa2** [comment]
test this please

**Comment by BillyAbildgaard** [comment]
+1

**Comment by BillyAbildgaard** [comment]
@yyscamper I made sure to address all of your comments.  I'll make sure to be more consistent with some of the styling going forward

**Pull request merged**